src/Client/MultiplexedConnections.h

```diff
@@ -74,12 +76,16 @@ class MultiplexedConnections final : public IConnections
     };
 
     /// Get a replica where you can read the data.
-    ReplicaState & getReplicaForReading();
+    ReplicaState & getReplicaForReading(size_t draining_timeout_seconds);
 
     /// Mark the replica as invalid.
     void invalidateReplica(ReplicaState & replica_state);
 
 private:
+    /// Hold a shared reference to the connection pool so that asynchronous connection draining will
+    /// work safely.
+    const ConnectionPoolWithFailoverPtr pool;
```

Comment:
It looks a little bit dangerous. If pool can own ConnectionPtr, and connection can own PoolPtr, we can get memory leak.
I suppose it does not happen for now, but I don't want to check (also, somebody could refactor code and not notice this).

Maybe we can use a pair `(pool, connection)` in `enqueueConnectionCleanup`?

Reply:
It's indeed entangled. I'll try refactoring the ownership.